library/payment/invoice.py

```diff
@@ -57,21 +58,21 @@ def calculate_fee(self, user: User) -> tuple[float, int]:
             reading_credits,
         )
 
-    def process_invoice_with_credit_card_detail(
-        self, number: str, cvv: str, expiration: datetime
-    ) -> bool:
-        card = CreditCard(number, expiration, cvv)
-        return self.process_invoice_with_credit_card(card)
-
-    def process_invoice_with_credit_card(self, card: CreditCard) -> bool:
+    def process_invoice(self, pay_method: Optional[CreditCard, Paypal]):
         if self.is_closed:
             # payment is already processed
             return True
         # validate card information
-        if not self._card_is_present_and_valid(card):
-            raise ValueError("Payment information is not set or not valid")
         fee, reading_credits = self.calculate_fee(self.customer)
-        is_paid: bool = self._pay_with_credit_card(card, fee)
+        if isinstance(pay_method, CreditCard):
```

Comment:
You could use the actual `CreditCard` and `PayPal` classes to process the invoices instead of doing an `isInstance` check.